client/src/main/java/io/pravega/client/stream/impl/LargeEventWriter.java

```diff
@@ -0,0 +1,252 @@
+import io.pravega.shared.protocol.netty.WireCommands.SegmentCreated;
+import io.pravega.shared.protocol.netty.WireCommands.SegmentIsSealed;
+import io.pravega.shared.protocol.netty.WireCommands.SegmentIsTruncated;
+import io.pravega.shared.protocol.netty.WireCommands.SegmentsMerged;
+import io.pravega.shared.protocol.netty.WireCommands.SetupAppend;
+import io.pravega.shared.protocol.netty.WireCommands.WrongHost;
+import java.nio.ByteBuffer;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.UUID;
+import java.util.concurrent.CompletableFuture;
+import lombok.Cleanup;
+import lombok.RequiredArgsConstructor;
+import lombok.val;
+import lombok.extern.slf4j.Slf4j;
+
+import static io.pravega.common.concurrent.Futures.getThrowingException;
+
+@RequiredArgsConstructor
+@Slf4j
+public class LargeEventWriter {
+
+    private static final int WRITE_SIZE = Serializer.MAX_EVENT_SIZE;
+    private final UUID writerId;
+    private final Controller controller;
+    private final ConnectionPool connectionPool;
+
+    public void writeLargeEvent(Segment segment, List<ByteBuffer> events, DelegationTokenProvider tokenProvider,
+            EventWriterConfig config) throws NoSuchSegmentException, AuthenticationException, SegmentSealedException {
+        List<ByteBuf> payloads = createBufs(events);
+        int attempts = 1 + Math.max(0, config.getRetryAttempts());
+        boolean sent = false;
+        Exception cause = null;
+        while (!sent && attempts > 0) {
+            try {
+                @Cleanup
+                RawClient client = new RawClient(controller, connectionPool, segment);
+                write(segment, payloads, client, tokenProvider);
+                sent = true;
+            } catch (ConnectionFailedException e) {
+                log.info("Connection failure while sending large event: {}. Retrying", e.getMessage());
+                cause = e;
+            } catch (TokenExpiredException e) {
+                tokenProvider.signalTokenExpired();
+                log.info("Authentication token expired while writing large event to segment {}. Retrying", segment);
+            }
+            attempts--;
+        }
+        if (!sent) {
+            throw new RetriesExhaustedException("Failed to write large event to segment " + segment, cause);
+        }
+    }
+
+    private List<ByteBuf> createBufs(List<ByteBuffer> events) {
+        ByteBuffer[] toWrite = new ByteBuffer[2 * events.size()];
+        for (int i = 0; i < events.size(); i++) {
+            ByteBuffer event = events.get(i);
+            byte[] header = new byte[WireCommands.TYPE_PLUS_LENGTH_SIZE];
+            ByteBuffer wrapped = ByteBuffer.wrap(header);
+            wrapped.putInt(WireCommandType.EVENT.getCode());
+            wrapped.putInt(event.remaining());
+            wrapped.flip();
+            toWrite[2 * i] = wrapped;
+            toWrite[2 * i + 1] = event;
+        }
+        ByteBuf master = Unpooled.wrappedBuffer(toWrite);
```

Comment:
`.wrappedUnmodifiableBuffer` is faster and returns a read-only buffer (which is what we want). Would that work here?

Reply:
That method only takes ByteBufs not ByteBuffers.

Reply:
That only works with ByteBufs and the incoming events are ByteBuffers so I would have to call `wrippedBuffer` on each of them individually first, which is surely worse than calling it once here.